pyof/v0x04/common/action.py

```diff
@@ -73,125 +75,214 @@ class ControllerMaxLen(IntEnum):
 # Classes
 
 
-class ActionExperimenterHeader(GenericStruct):
-    """Action structure for OFPAT_EXPERIMENTER."""
+class ActionHeader(GenericStruct):
+    """Action header that is common to all actions.
+
+    The length includes the header and any padding used to make the action
+    64-bit aligned.
+    NB: The length of an action *must* always be a multiple of eight.
+    """
 
-    #: OFPAT_EXPERIMENTER.
-    action_type = UBInt16(ActionType.OFPAT_EXPERIMENTER, enum_ref=ActionType)
-    #: Length is multiple of 8.
+    #: One of OFPAT_*.
+    action_type = UBInt16(enum_ref=ActionType)
+    #: Length of action, including this header. This is the length of actions,
+    #:    including any padding to make it 64-bit aligned.
     length = UBInt16()
-    #: Experimenter ID which takes the same form as in struct
-    #:     ofp_experimenter_header
+    # Pad for 64-bit alignment.
+    # This should not be implemented, as each action type has its own padding.
+    # pad = Pad(4)
+
+    _allowed_types = ()
+
+    def __init__(self, action_type=None, length=None):
+        """Create an ActionHeader with the optional parameters below.
+
+        Args:
+            action_type (~pyof.v0x04.common.action.ActionType):
+                The type of the action.
+            length (int): Length of action, including this header.
+        """
+        super().__init__()
+        self.action_type = action_type
+        self.length = length
+
+    def unpack(self, buff, offset=0):
+        """Unpack a binary message into this object's attributes.
+
+        Unpack the binary value *buff* and update this object attributes based
+        on the results.
+
+        Args:
+            buff (bytes): Binary data package to be unpacked.
+            offset (int): Where to begin unpacking.
+
+        Raises:
+            Exception: If there is a struct unpacking error.
+
+        """
+        self.action_type = UBInt16(enum_ref=ActionType)
+        self.action_type.unpack(buff, offset)
+
+        for cls in ActionHeader.__subclasses__():
+            if self.action_type.value in cls.get_allowed_types():
```

Comment:
This **if** (`if self.action_type.value in cls.get_allowed_types():`) statement will not work in all cases.

In many cases, the `_allowed_types` are just a class, not a list. Thus the `in` operator will raise an exception:
```
In [1]: from pyof.v0x01.common.action import *

In [2]: a = ActionOutput

In [3]: ActionOutput in a
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-4e99ae1f96d2> in <module>()
----> 1 ActionOutput in a

TypeError: argument of type 'MetaStruct' is not iterable
```
or
```
In [6]: a = ActionType.OFPAT_OUTPUT

In [7]: 0 in a
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-b91a8fa869f1> in <module>()
----> 1 0 in a

TypeError: argument of type 'int' is not iterable
```
We have two options here to make this conditional work:

1. Ensure that `_allowed_types` is always an iterable (a `list` or a `set`);
2. Change the if statement to:
`if self.action_type.value == cls.get_allowed_types() or self.action_type.value in cls.get_allowed_types():`

But it may be a better approach to create a `classmethod` that do the check. So you can just use it as `if cls.is_type_allowed(self.action_type):` or something like that.

Reply:
I have changed them to iterables to match the way it's implemented in v0x01.